README.md

# 광운대학교_소프트웨어학부_졸업작품 <br> "Raspberry Pi를 활용한 스마트 거울"

## 🤷‍♂️ 무슨 프로젝트 인가요?

**➡️** 라즈베리파이를 활용하여 거울의 기본적인 기능을 함은 물론이고, Google Assistant, 안면인식을 통한 개인 맞춤 알림 서비스, Azure를 활용한 온/습도 데이터화등 심화된 IoT 기능을 제공하도록 구성한 프로젝트입니다.

## 💁‍♂️ 어떤 식으로 구성되고 작동하나요?

**➡️** 구성 방식은 다음과 같습니다.

### HW

**- 전면**

![image](https://user-images.githubusercontent.com/65067460/140688918-db1c88a2-63f8-413f-bb2a-cae63f511393.png)


**- 후면**

![image](https://user-images.githubusercontent.com/65067460/140688831-6276e7cf-ac68-4a8e-97e6-d4674d06ca05.png)

### SW

![image](https://user-images.githubusercontent.com/65067460/140689253-7b0a83e2-8beb-4da4-a614-f3da18ccb41e.png)


**- MagicMirror 2**
https://github.com/MichMich/MagicMirror

**- MagicMirroe 2 Modules (Google Assistant, Face Recognition, Phone Notification 등)**
https://github.com/MichMich/MagicMirror/wiki/3rd-Party-Modules

**- MicroSoft Azure IoT**
https://docs.microsoft.com/ko-kr/azure/iot-hub/iot-hub-raspberry-pi-kit-node-get-started

**- Grafana를 활용한 온/습도 시계열 데이터화**
https://sandervandevelde.wordpress.com/2021/06/15/cloud-iot-dashboards-using-grafana-with-azure-iot/


## 🧐 어떠한 기능이 있죠?

➡️ 구성된 기능은 다음과 같습니다.

|기능|상세 소개|자세히 알아보기| 
|:---:|:---:|:---| 
|시계, 날씨, 일정|제공되는 기본 기능|https://docs.magicmirror.builders/development/node-helper.html#module-instance-methods| 
|유튜브 스트리밍|유튜브 N-Screen 기능||
|휴대전화 Notification 연동|개인 휴대전화의 알림을 거울로 확인|https://github.com/basknol/MMM-PushBulletNotifications|
|안면인식|안면인식을 기능|https://github.com/normyx/MMM-Facial-Recognition-OCV3|
|프로필 체인저|'안면인식' 기능과 연동, 개인 맞춤 서비스|https://github.com/tosti007/MMM-ProfileSwitcher|
|Google Assistant|'구글 어시스턴트' 기능, 각종 'Action'을 통해 다양한 음성인식 솔루션 이용 가능|http://wiki.bugsounet.fr/en/MMM-GoogleAssistant|
|온/습도 빅데이터화|Azure와 Grafana를 활용, 온/습도 데이터를 빅데이터화 함|https://sandervandevelde.wordpress.com/2021/06/15/cloud-iot-dashboards-using-grafana-with-azure-iot/|

# 위 기능을 구현하기 위해 다음의 프로젝트를 사용하였습니다.




For the full documentation including **[installation instructions](https://docs.magicmirror.builders/getting-started/installation.html)**, please visit our dedicated documentation website: [https://docs.magicmirror.builders](https://docs.magicmirror.builders).





<p style="text-align: center">
	<a href="https://bestpractices.coreinfrastructure.org/projects/347"><img src="https://bestpractices.coreinfrastructure.org/projects/347/badge" alt="CLI Best Practices"></a>
	<a href="https://codecov.io/gh/MichMich/MagicMirror"><img src="https://codecov.io/gh/MichMich/MagicMirror/branch/master/graph/badge.svg?token=LEG1KitZR6" alt="CodeCov Status"/></a>
	<a href="https://choosealicense.com/licenses/mit"><img src="https://img.shields.io/badge/license-MIT-blue.svg" alt="License"></a>
	<a href="https://github.com/MichMich/MagicMirror/actions?query=workflow%3A%22Automated+Tests%22"><img src="https://github.com/MichMich/MagicMirror/workflows/Automated%20Tests/badge.svg" alt="Tests"></a>
</p>

**MagicMirror²** is an open source modular smart mirror platform. With a growing list of installable modules, the **MagicMirror²** allows you to convert your hallway or bathroom mirror into your personal assistant. **MagicMirror²** is built by the creator of [the original MagicMirror](https://michaelteeuw.nl/tagged/magicmirror) with the incredible help of a [growing community of contributors](https://github.com/MichMich/MagicMirror/graphs/contributors).

MagicMirror² focuses on a modular plugin system and uses [Electron](https://www.electronjs.org/) as an application wrapper. So no more web server or browser installs necessary!

## Documentation

For the full documentation including **[installation instructions](https://docs.magicmirror.builders/getting-started/installation.html)**, please visit our dedicated documentation website: [https://docs.magicmirror.builders](https://docs.magicmirror.builders).

## Links

- Website: [https://magicmirror.builders](https://magicmirror.builders)
- Documentation: [https://docs.magicmirror.builders](https://docs.magicmirror.builders)
- Forum: [https://forum.magicmirror.builders](https://forum.magicmirror.builders)
  - Technical discussions: https://forum.magicmirror.builders/category/11/core-system
- Discord: [https://discord.gg/J5BAtvx](https://discord.gg/J5BAtvx)
- Blog: [https://michaelteeuw.nl/tagged/magicmirror](https://michaelteeuw.nl/tagged/magicmirror)
- Donations: [https://magicmirror.builders/#donate](https://magicmirror.builders/#donate)

## Contributing Guidelines

Contributions of all kinds are welcome, not only in the form of code but also with regards to

- bug reports
- documentation
- translations

For the full contribution guidelines, check out: [https://docs.magicmirror.builders/getting-started/contributing.html](https://docs.magicmirror.builders/getting-started/contributing.html)

## Enjoying MagicMirror? Consider a donation!

MagicMirror² is opensource and free. That doesn't mean we don't need any money.

Please consider a donation to help us cover the ongoing costs like webservers and email services.
If we receive enough donations we might even be able to free up some working hours and spend some extra time improving the MagicMirror² core.

To donate, please follow [this](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=G5D8E9MR5DTD2&source=url) link.

<p style="text-align: center">
	<a href="https://forum.magicmirror.builders/topic/728/magicmirror-is-voted-number-1-in-the-magpi-top-50"><img src="https://magicmirror.builders/img/magpi-best-watermark-custom.png" width="150" alt="MagPi Top 50"></a>
</p>


CHANGELOG.md
# MagicMirror² Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](https://semver.org/).

❤️ **Donate:** Enjoying MagicMirror²? [Please consider a donation!](https://magicmirror.builders/donate) With your help we can continue to improve the MagicMirror²

## [2.17.1] - 2021-10-01

### Fixed

- Fixed error when accessing letsencrypt certificates

## [2.17.0] - 2021-10-01

Special thanks to the following contributors: @apiontek, @eouia, @jupadin, @khassel and @rejas.

### Added

- Added showTime parameter to clock module for enabling/disabling time display in analog clock.
- Added custom electron switches from user config (`config.electronSwitches`).
- Added unit tests for updatenotification module.

### Updated

- Bump electron to v13 (and spectron to v15) and update other dependencies in package.json.
- Refactor test configs, use default test config for all tests.
- Updated github templates.
- Actually test all js and css files when lint script is run.
- Update jsdocs and print warnings during testing too.
- Update weathergov provider to try fetching not just current, but also foreacst, when API URLs available.
- Refactored clock layout.
- Refactored methods from weatherproviders into weatherobject (isDaytime, updateSunTime).
- Use of `logger.js` in jest tests.
- Run prettier over all relevant files.
- Move tests needing electron in new category `electron`, use `server only` mode in `e2e` tests.
- Update dependencies in package.json.

### Fixed

- Fix undefined error with ignoreToday option in weather module (#2620).
- Fix time zone correction in calendar module when the date hour is equal to the time zone correction value (#2632).
- Fix black cursor on startup when using electron.
- Fix update notification not working for own repository (#2644).

## [2.16.0] - 2021-07-01

Special thanks to the following contributors: @210954, @B1gG, @codac, @Crazylegstoo, @daniel, @earlman, @ezeholz, @FrancoisRmn, @jupadin, @khassel, @KristjanESPERANTO, @njwilliams, @oemel09, @r3wald, @rejas, @rico24, Faizan Ahmed.

### Added

- Added French translations for "MODULE_CONFIG_ERROR" and "PRECIP".
- Added German translation for "PRECIP".
- Added Dutch translation for "WEEK", "PRECIP", "MODULE_CONFIG_CHANGED" and "MODULE_CONFIG_ERROR".
- Added first test for Alert module.
- Added support for `dateFormat` when not using `timeFormat: "absolute"`.
- Added custom-properties for colors and fonts for improved styling experience, see `custom.css.sample` file.
- Added custom-properties for gaps around body and between modules.
- Added test case for recurring calendar events.
- Added new Environment Canada provider for default WEATHER module (weather data for Canadian locations only).
- Added list view for newsfeed module.
- Added dev dependency jest, switching from mocha to jest.

### Updated

- Bump node-ical to v0.13.0 (now last runtime dependency using deprecated `request` package is removed).
- Use codecov in informational mode.
- Refactor code into es6 where possible (e.g. var -> let/const).
- Use node v16 in github workflow (replacing node v10).
- Moved some files into better suited directories.
- Update dependencies in package.json, require node >= v12, remove `rrule-alt` and `rrule`.
- Update dependencies in package.json and migrate husky to v6, fix husky setup in prod environment.
- Cleaned up error handling in newsfeed and calendar modules for real.
- Updated default WEATHER module such that a provider can optionally set a custom unit-of-measure for precipitation (`weatherObject.precipitationUnits`).
- Update documentation.
- Update jest tests: Reset changes on js/logger.js, mock logger.js in global_vars tests.
- Update dependencies in package.json.

### Removed

- Switching from mocha to jest so removed following dev dependencies: chai, chai-as-promised, mocha, mocha-each, mocha-logger.

### Fixed

- Fix calendar start function logging inconsistency.
- Fix updatenotification start function logging inconsistency.
- Checks and applies the showDescription setting for the newsfeed module again.
- Fix issue with openweathermap not showing current or forecast info when using onecall API.
- Fix tests in weather module and add one for decimalPoint in forecast.
- Fix decimalSymbol in the forecast part of the new weather module (#2530).
- Fix wrong treatment of `appendLocationNameToHeader` when using `ukmetofficedatahub`.
- Fix alert not recognizing multiple alerts (#2522).
- Fix fetch option httpsAgent to agent in calendar module (#466).
- Fix module updatenotification which did not work for repos with many refs (#1907).
- Fix config check failing when encountering let syntax ("Parsing error: Unexpected token config").
- Fix calendar debug check.
- Really run prettier over all files.
- Fix logger.js after jest changes, use --forceExit running jest.
- Workaround for dev_console test using getWindowCount.

## [2.15.0] - 2021-04-01

Special thanks to the following contributors: @EdgardosReis, @MystaraTheGreat, @TheDuffman85, @ashishtank, @buxxi, @codac, @fewieden, @khassel, @klaernie, @qu1que, @rejas, @sdetweil & @thomasrockhu.

ℹ️ **Note:** This update uses new dependencies. Please update using the following command: `git pull && npm install`.

### Added

- Added Galician language.
- Added GitHub workflows for automated testing and changelog enforcement.
- Added CodeCov badge to Readme.
- Added CURRENTWEATHER_TYPE notification to currentweather and weather module, use it in compliments module.
- Added `start:dev` command to the npm scripts for starting electron with devTools open.
- Added logging when using deprecated modules weatherforecast or currentweather.
- Added Portuguese translations for "MODULE_CONFIG_CHANGED" and "PRECIP".
- Respect parameter ColoredSymbolOnly also for custom events.

config/config.js
let config = {
	address: "localhost", 	
	port: 8080,
	basePath: "/", 	
	ipWhitelist: ["127.0.0.1", "::ffff:127.0.0.1", "::1"], 	

	useHttps: false, 		
	httpsPrivateKey: "", 	
	httpsCertificate: "",

	language: "en",
	locale: "en-US",
	logLevel: ["INFO", "LOG", "WARN", "ERROR"], 
	timeFormat: 24,
	units: "metric",

	modules: [
		{
			module: "alert",
			classes: 'pill kimtree'
		},
		{
			module: "updatenotification",
			position: "top_bar",
			classes: 'pill kimtree'
		},
		{
			module: "clock",
			position: "top_left",
			classes: 'everyone'
		},
		{
		module: 'MMM-Screencast',
		classes: 'everyone',
		position: 'top_left',
		config: {
			position: 'bottomCenter',
			height: 150,
			width: 300,
		}
        },
		{
			module: "calendar2",
			header: "내 저장된 일정",
			position: "top_left",
			classes: 'pill',
			config: {
				calendars: [
					{
						symbol: "calendar-check",
						
						url: "https://calendar.google.com/calendar/ical/pansyscent%40gmail.com/private-f974f2b13ddc006bd7a5b517cc4f1462/basic.ics"	}
				]
			}
		},
				{
			module: "calendar",
			header: "내 저장된 일정",
			position: "top_left",
			classes: 'kimtree',
			config: {
				calendars: [
					{
						symbol: "calendar-check",
						
						url: "https://calendar.google.com/calendar/ical/nawkim98%40gmail.com/public/basic.ics"					
						}
				]
			}},
		{
			module: 'MMM-BackgroundSlideshow',
			classes:'default',
			position: 'fullscreen_below',
			config: {
			imagePaths: ['modules/MMM-BackgroundSlideshow/exampleImages/'],
			transitionImages: false,
			randomizeImageOrder: true
		}
		},

		{
			module: "weather",
			position: "top_right",
			header: "현재 날씨",
			classes: 'pill kimtree',
			config: {
				weatherProvider: "openweathermap",
				type: "current",
				location: "서울특별시",
				locationID: "1835847",
				apiKey: "9672c481ee3879ad9bb2e3166974367f"
			}
		},
		{
			module: "weather",
			position: "top_right",
			header: "날씨 예보",
			classes: 'pill kimtree',
			config: {
				weatherProvider: "openweathermap",
				type: "forecast",
				location: "서울특별시",
				locationID: "1835847",
				apiKey: "9672c481ee3879ad9bb2e3166974367f"
			}
		},
		{
			module: "newsfeed",
			position: "bottom_bar",
			classes: 'pill kimtree',
			config: {
				feeds: [
					{
						title: "현재 주요 뉴스",
						url: "https://news.google.com/rss?gl=KR&hl=ko&ceid=KR:ko"
					}
				],
				showSourceTitle: true,
				showPublishDate: true,
				broadcastNewsFeeds: true,
				broadcastNewsUpdates: true
			}
		},
		
		{
			module: 'MMM-AirQuality',
			position: 'top_left',
			classes: 'pill kimtree',
			config: {
			location: 'seoul',
			lang: "kr"
			}
		},
		{
			module: 'MMM-Face-Reco-DNN',
			config: {
			 
			  logoutDelay:60000,
			  checkInterval: 2000,
			  noFaceClass: 'noface',
			  unknownClass: 'unknown',
			  knownClass: 'known',
			  defaultClass: 'default',
			  everyoneClass: 'everyone',
			  alwaysClass: 'always',
			  cascade: 'modules/MMM-Face-Reco-DNN/tools/haarcascade_frontalface_default.xml',
			  encodings: 'modules/MMM-Face-Reco-DNN/tools/encodings.pickle',
			  usePiCamera: 1,
			  source: 0,
			  rotateCamera: 0,
			  method: 'dnn',
			  detectionMethod: 'hog',
			  animationSpeed: 0,
			  pythonPath: null,
			  welcomeMessage: true,
			  usernameDisplayMapping: null,
			  extendDataset: false,
			  dataset: 'modules/MMM-Face-Reco-DNN/dataset/',
			  tolerance: 0.6,
			  multiUser: 0,
			}
		},

		

		{
			 module: "MMM-DHT-Sensor",
			 position: 'top_right',
			 header: "실내 온습도",
			 classes: 'pill kimtree',
			 config: {
				sensorPin: 4,
				sensorType: 11,
				updateInterval: 1 * 1000,
				initialLoadDelay: 0,
				animationSpeed: 1000,
				units: "metric",
				relativeScale: 35,
				debug: false
				}
		},
		/*{
		module: 'MMM-iFrame',
		classes: 'pill kimtree',
		position: 'top_right',	// This can be any of the regions.
		config: {
			// See 'Configuration options' for more information.
				url: ["http://52.231.91.247:3000/d/dM3j6fFnk/smartmirror_ai_temperature?orgId=1&viewPanel=2&refresh=1d&from=1636008855947&to=1636095255948"],  // as many URLs you want or you can just ["ENTER IN URL"] if single URL.
				updateInterval: 0.5 * 60 * 1000, // rotate URLs every 30 seconds
				width: "400", // width of iframe
				height: "700", // height of iframe
				frameWidth: "600" // width of embedded iframe, height is beeing calculated by aspect ratio of iframe
			}
		},*/
		{ 
			module : 'MMM-GrafanaChart', 
		    position : 'top_right' ,  
		    header : "실내 온습도 그래프",
			classes: 'pill kimtree', 
			config : { 
					version : "6" ,	 
					id : "dM3j6fFnk" , 
					host : "52.231.91.247" , 
					port: 3000,
					dashboardname: "smartmirror_ai_temperature", 
					orgId: 1, 
					panelId: 2,
					width: "100%", 
					height: "100%",
					scrolling: "yes",
					refreshInterval: 900
				}
		}, 
		/*{
	 module: 'MMM-GrafanaGauges',
		 position: 'top_right',   // This can be any of the regions.
		 classes: 'pill kimtree',
         header: 'Test Graph',
		 config: {
					host: "20.196.219.156", //Mandatory. See url when displaying within grafana
					port: 3000, // Mandatory.
					dashboardname: "smartmirror_ai_temperature", // Mandatory.  
					orgId: 1, // Mandatory.
					showIDs: [2],// Mandatory. PanelId from the url.  
					width: "100%", // Optional. Default: 100%
					height: "100%", // Optional. Default: 100%
					refreshInterval: 10 //Optional. Default: 900 = 1/4 hour
				}
},*/
		{
		module: 'MMM-PushBulletNotifications',
		header: 'Notifications',
		classes: "kimtree",
		position: 'top_left',
		config: {
			accessToken: "o.9hehtgGPGqkeVdoa2atfcQVvWEORWjBB",
			endToEndPassword: null,
			numberOfNotifications: 3,
			filterTargetDeviceName: "",
			showPushesSentToAllDevices: true,
			onlyAllowCommandsFromSourceDevices: [],
			fetchLimitPushBullet: 50,
			showPushes: true,
			showPushesOnLoad: true,
			showDismissedPushes: true,
			showMirroredNotifications: true, 
			onlyShowLastNotificationFromApplication: false,
			showIndividualNotifications: false,
			showSMS: true,
			showMessage: true,
			showIcons: true,
			showDateTime: true,
			localesDateTime: 'nl-NL',
			playSoundOnNotificationReceived: true,
			soundFile: 'modules/MMM-PushBulletNotifications/sounds/new-message.mp3',			
			maxMsgCharacters: 50,
			maxHeaderCharacters: 32,
			showModuleIfNoNotifications: true,
			noNotificationsMessage: "새 메세지가 없습니다.",
			debugMode: false,
		}
		},
				{
		module: 'MMM-PushBulletNotifications2',
		header: 'Notifications',
		classes: "pill",
		position: 'top_left',
		config: {
			accessToken: "o.i3i8kgQwwf0cKYF86pUjGUDzojQTmXrc",
			endToEndPassword: null,
			numberOfNotifications: 3,
			filterTargetDeviceName: "",
			showPushesSentToAllDevices: true,
			onlyAllowCommandsFromSourceDevices: [],
			fetchLimitPushBullet: 50,
			showPushes: true,
			showPushesOnLoad: true,
			showDismissedPushes: true,
			showMirroredNotifications: true, 
			onlyShowLastNotificationFromApplication: false,
			showIndividualNotifications: false,
			showSMS: true,
			showMessage: true,
			showIcons: true,
			showDateTime: true,
			localesDateTime: 'nl-NL',
			playSoundOnNotificationReceived: true,
			soundFile: 'modules/MMM-PushBulletNotifications2/sounds/new-message.mp3',			
			maxMsgCharacters: 50,
			maxHeaderCharacters: 32,
			showModuleIfNoNotifications: true,
			noNotificationsMessage: "새 메세지가 없습니다.",
			debugMode: false,
		}
		},
		{
			module: "MMM-MovieListings",
			position: "top_left",
			classes: 'pill kimtree',
			config: {
				apiKey : "e7627842dd1b7a490f26016e1e659316",
				header: "상영중인 영화",
				region : 'KR',
				lanquage : 'ko-KR',
				pageChangeInterval : 10*1000,
				
			}
			},



				{
		module: 'MMM-iFrame',
		position: 'top_right',
		classes:'pill kimtree',
		config: {
				url: ["http://localhost:3000/"],
				updateInterval: 0.5 * 60 * 10000,
				width: "100",
				height: "100",
				frameWidth: "500"
			}
		},
		{
		  module: "MMM-GoogleAssistant",
		  position: "top_left",
		  classes: 'default kimtree pill',
		  configDeepMerge: true,
		  config: {
			assistantConfig: {
			  latitude: 51.508530,
			  longitude: -0.076132,
			},
		  }
		},
		{
		  module: "MMM-Detector",
		  configDeepMerge: true,
		  classes: 'default kimtree pill',
		  config: {
			debug: false,
			autoStart: true,
			micConfig: {
			  recorder: "arecord",
			  device: "plughw:2",
			  audioGain: 2.0,
			  applyFrontend: true,
			},
			detectors: [
			    {
				  detector: "Snowboy",
				  Model: "jarvis",
				  Sensitivity: null,
				  Logo: "google",
				  autoRestart: false,
				  onDetected: {
					notification: "GA_ACTIVATE"
				  }
				}
			
			  

			]
		  }
		}
	

		
	]
};

/*************** DO NOT EDIT THE LINE BELOW ***************/
if (typeof module !== "undefined") {module.exports = config;}
